tests/test_updater.py

```diff
@@ -331,6 +330,45 @@ def attempt(*args, **kwargs):
             updater._bootstrap(retries, False, 'path', None, bootstrap_interval=0)
         assert self.attempts == attempts
 
+    def test_bootstrap_clean_updates(self, monkeypatch, updater):
+        clean = True
+        expected_id = 4
+        self.offset = 0
+
+        def get_updates(*args, **kwargs):
+            # we're hitting this func twice
+            # 1. no args, return list of updates
+            # 2. with 1 arg, int => if int == expected_id => test successful
+
+            # case 2
+            # 2nd call from bootstrap____clean
+            # we should be called with offset = 4
+            # save value passed in self.update_id for assert down below
+            if len(args) > 0:
+                self.offset = int(args[0])
+                return []
```

Comment:
Maybe change is to 

```python
return [FakeUpdate(i) for i in range(args[0], expected_id)]
```

and add `assert bot.get_updates == []` at the very end.

Reply:
TL;DR: changing this line as suggested doesn't work and break current test-strategy. I've pushed commits to make this work but needed to rearrange quite some stuff and store the list of updates in self.updates


if we want to assert bot.get_updates() = [] than we will call the fake get_updates for a 3rd time. It's called w/o an argument so we'll fall through case 2 and go down to case 1 and return a list of fake objects.

If you want to do this we need to store the FakeUpdate list in self.updates outside the fake get_updates, before the first call, to ensure it's available for (every) call to the fake get_updates. If called we will return self.updates with 3 fake updates (1st call). We than also need to have the self.updates list modified when get_updates is called with the offset (2nd call, with arg). After all that we can return self.updates on assert (3rd call).

Reply:
ps: please not that this
`return [FakeUpdate(i) for i in range(args[0], expected_id)]
`
returns [] when args[0] > expected_id. This should raise an error in my opinion

Reply:
Okay. I think the `expected_id` test is more important than testing `get_updates() = []` afterwards (as that's actually something TG promises), just thought it would be a minimal-effort addition. Guess I was wrong … Please revert. Then we're good to go :)

Reply:
I think this is it than?